mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp

```diff
@@ -291,26 +300,30 @@ static bool hasTensorSignature(func::FuncOp funcOp) {
 /// retrieve the called FuncOp from any func::CallOp.
 static LogicalResult
 getFuncOpsOrderedByCalls(ModuleOp moduleOp,
-                         SmallVectorImpl<func::FuncOp> &orderedFuncOps,
+                         SmallVectorImpl<FunctionOpInterface> &orderedFuncOps,
                          FuncCallerMap &callerMap) {
   // For each FuncOp, the set of functions called by it (i.e. the union of
   // symbols of all nested func::CallOp).
-  DenseMap<func::FuncOp, DenseSet<func::FuncOp>> calledBy;
+  DenseMap<FunctionOpInterface, DenseSet<FunctionOpInterface>> calledBy;
   // For each FuncOp, the number of func::CallOp it contains.
-  DenseMap<func::FuncOp, unsigned> numberCallOpsContainedInFuncOp;
-  WalkResult res = moduleOp.walk([&](func::FuncOp funcOp) -> WalkResult {
-    if (!funcOp.getBody().empty()) {
-      func::ReturnOp returnOp = getAssumedUniqueReturnOp(funcOp);
-      if (!returnOp)
-        return funcOp->emitError()
-               << "cannot bufferize a FuncOp with tensors and "
-                  "without a unique ReturnOp";
+  DenseMap<FunctionOpInterface, unsigned> numberCallOpsContainedInFuncOp;
+  WalkResult res = moduleOp.walk([&](FunctionOpInterface funcOp) -> WalkResult {
+    // Only handle ReturnOp if funcOp is exactly the FuncOp type.
+    if(isa<FuncOp>(funcOp)) {
```

Comment:
I think you can count the number of `ReturnLike` (that's a trait) operations in the the region of the `funcOp`. (Not in nested regions.)

Reply:
Hi, we have made `getAssumedUniqueReturnOp` to detect `ReturnLike` ops and made it return `funcOp` if there is no `ReturnLike` op. When it returns `funcOp` ifselt, the analysis in `aliasingFuncOpBBArgsAnalysis` and `foldMemRefCasts` will be skipped.

In summary, we think there are 3 scenarios:
- Exaclty one `ReturnLike` op => Perform the return-related analysis.
- No `ReturnLike` op  => Skip the analysis. No error is thrown.
- Multiple `ReturnLike` ops => Throw an error.

Reply:
@tzunghanjuang is this because the functions have no body (i.e., they are just a declaration)? Or do we have an terminator op that does not implement the ReturnLike trait that is triggering this?

Reply:
For a bit more context, we think there are some `FunctionOpInterface`s ,like [`transform.NamedSequenceOp`](https://github.com/llvm/llvm-project/blob/109cd11dc4aea6b3596f8b2cb5a719f35b190cfa/mlir/include/mlir/Dialect/Transform/IR/TransformOps.td#L993-L997), having no `ReturnLike` at the end.

Example: https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Vector/transform-vector.mlir

Reply:
I would throw an error even if there no `ReturnLike` op, and relax this requirement only if needed. `transform.named_sequence` does not return tensors, so we wouldn't bufferize it in the first place.

Reply:
Thank you for the suggestion. We have updated `getAssumedUniqueReturnOp` so that it no longer returns `funcOp`. We did not find a directly determine if a `FunctionOpInterface` is shipped with any `ReturnLike` Op, so we used  `funcOp.getNumResults()` to check if there is no return type. 

With this method, we still have to keep the check near `aliasingFuncOpBBArgsAnalysis` and `foldMemRefCasts` or the following example will fail. The example uses `FuncOP` and returns nothing. If we check `funcOp.getNumResults()` too early, the bufferization of the function's body will be entirely skipped and fail the file check. 

https://github.com/llvm/llvm-project/blob/fc7a8936208847b66df9b5f93ba2222b93b9f8b6/mlir/test/Dialect/ControlFlow/one-shot-bufferize-analysis.mlir#L91-L111

Reply:
Hi @tzunghanjuang, you should not use `getNumResults` in this manner:

```
    if (!funcOp.getFunctionBody().empty() && funcOp.getNumResults() != 0) {
```

One of the errors appears to be in the tests themselves. I'll explain here for @matthias-springer:

The tests that involve the transform interpreter will attempt to bufferize the root module which contains the `transform::NamedSequenceOp`. The `transform::NamedSequenceOp` implements `FunctionOpInterface`, but the dialect does not have a `ReturnLike` op. The one-shot bufferizer will attempt to bufferize all nested operations inside the root module which includes the `transform::NamedSequenceOp`. However, this will obviously fail.

So I believe the correct way to solve this is to change the test such that the bufferization only happens in the intended sections of the module. This looks like it can be achieved by specifying a payload target with the option `--debug-payload-root-tag=<string>` or (as I just looked into other tests) using the `transform.structured.match` operation.

Looking at the tests, it seems that `transform.structured.match` is preferred, but I'll let others chime in.

Reply:
Thank you for the clarification. I have updated the `transform.named_sequence` tests with `-debug-payload-root-tag` option. The previous workaround has also been removed. The "no ReturnLike op" cases will throw errors.